planex/cmd/pin.py

```diff
@@ -0,0 +1,233 @@
+    if repo.url is not None:
+        return urlunparse(repo.url)
+
+
+def commitish_tag_or_branch(repo):
+    """
+    Return the first non-null value among the repository commitish,
+    tag or branch. If they are all None, it returns None.
+    """
+    if repo.commitish is not None:
+        return repo.commitish
+    if repo.tag is not None:
+        return repo.tag
+    if repo.branch is not None:
+        return repo.branch
+
+
+def get_pin_content(args, pq_name, spec, link):
+    """
+    Generate the pinfile content for a Spec.
+    """
+    base_repo = repository_of(spec)
+    pq_repo = repository_of(link)
+
+    if args.url is not None:
+        url = args.url
+    else:
+        if link is not None:
+            url = repository_url(pq_repo)
+        else:
+            url = repository_url(base_repo)
+
+    if args.branch is not None:
+        commitish = args.branch
+    else:
+        if link is not None:
+            # A link specifies a tag, version or commitish
+            commitish = commitish_tag_or_branch(pq_repo)
+        else:
+            commitish = "master"
+
+    pinfile = {
+        'URL': url,
+        'commitish': commitish,
+        'patchqueue': pq_name
+    }
+
+    if link is not None:
+        if args.base is not None:
+            base_url, base_commitish = args.base.split("#", 1)
+            base = base_url
+            base_commitish = base_commitish
+        else:
+            base = repository_url(base_repo)
+            # TODO: what whould we do if base is a url to a non-git tarball and the following is None?
+            base_commitish = commitish_tag_or_branch(base_repo)
+
+        # TODO: Should we take into account here if a spec is already in PINS?
+        pinfile.update({
+            'base': base,
+            'base_commitish': base_commitish
+        })
+
+        if link.sources is not None:
+            pinfile["sources"] = link.sources
+
+        if link.patches is not None:
+            pinfile["patches"] = link.patches
+
+    return pinfile
+
+
+def make_pin(args, xs_path, package_name):
+    """
+    Return the pinfile path and a dict containing the content of the pinfile
+    """
+
+    # TODO: make the 'patchqueue' field customisable?
+    pin_pq_name = "master"
+    spec, link = spec_and_lnk(xs_path, package_name)
+
+    if link is None and args.base is not None:
+        print("Argument --base is allowed only for lnk packages")
+        sys.exit(1)
+
+    return get_pin_content(args, pin_pq_name, spec, link)
+
+
+def parse_args_or_exit(argv=None):
+    """
+    Parse command line options
+    """
+
+    parser = argparse.ArgumentParser(
+        description="Create a .pin file pointing to a repository "
+                    "in $CWD/repos. You must run "
+                    "this tool from the root of a spec repository.")
+    add_common_parser_options(parser)
+    parser.add_argument("package", help="package name")
+    parser.add_argument("--pinsdir", default="PINS",
```

Comment:
I think we should print the new pin to stdout and let the user redirect it to the appropriate file.   If we do that, then we can get rid of the `--pinsdir` and `--dry-run` options and the code that goes with them.